src/overlay/SurveyDataManager.h

```diff
@@ -0,0 +1,206 @@
+#include "util/Timer.h"
+#include "util/UnorderedMap.h"
+
+#include "medida/meter.h"
+#include "xdr/Stellar-overlay.h"
+#include "xdr/Stellar-types.h"
+
+#include <chrono>
+#include <mutex>
+#include <optional>
+
+namespace stellar
+{
+
+enum class SurveyPhase
+{
+    // Survey is currently recording data
+    RECORDING,
+    // Recording complete. Survey data is available for reporting.
+    REPORTING,
+    // No active survey in progress. No data is being recording or reported.
+    INACTIVE
+};
+
+struct RecordingNodeData
+{
+    RecordingNodeData(uint64_t initialLostSyncCount);
+
+    // Peer change data
+    uint32_t mAddedAuthenticatedPeers = 0;
+    uint32_t mDroppedAuthenticatedPeers = 0;
+
+    // SCP stats
+    std::vector<std::chrono::nanoseconds> mSCPFirstToSelfLatencies;
+    std::vector<std::chrono::nanoseconds> mSCPSelfToOtherLatencies;
+
+    // To compute how many times the node lost sync in the time slice
+    uint64_t const mInitialLostSyncCount;
+};
+
+// Data about a peer
+struct RecordingPeerData
+{
+    RecordingPeerData(Peer::PeerMetrics const& peerMetrics);
+
+    // Metrics at the start of the survey
+    uint64_t const mInitialMessageRead;
+    uint64_t const mInitialMessageWrite;
+    uint64_t const mInitialByteRead;
+    uint64_t const mInitialByteWrite;
+    uint64_t const mInitialUniqueFloodBytesRecv;
+    uint64_t const mInitialDuplicateFloodBytesRecv;
+    uint64_t const mInitialUniqueFetchBytesRecv;
+    uint64_t const mInitialDuplicateFetchBytesRecv;
+    uint64_t const mInitialUniqueFloodMessageRecv;
+    uint64_t const mInitialDuplicateFloodMessageRecv;
+    uint64_t const mInitialUniqueFetchMessageRecv;
+    uint64_t const mInitialDuplicateFetchMessageRecv;
+
+    // For computing average latency (in miliseconds)
+    std::chrono::milliseconds mLatencySum{0};
+    uint64_t mLatencyCount = 0;
+};
+
+/*
+ * Manage data collection during time sliced overlay surveys. This class is
+ * thread-safe.
+ */
+class SurveyDataManager : public NonMovableOrCopyable
+{
+  public:
+    // Create a survey manager. `clock` must be the Application's clock.
+    // `lostSyncMeter` is a meter to track how many times the node lost sync.
+    SurveyDataManager(VirtualClock const& clock,
+                      medida::Meter const& lostSyncMeter);
+
+    // Start the recording phase of a survey. Ignores requests if a survey is
+    // already active. `peers` should contain the `NodeID`s of all connected
+    // peers. Returns `true` if this successfully starts a survey.
+    bool startSurveyRecording(TimeSlicedSurveyStartRecordingMessage const& msg,
+                              std::map<NodeID, Peer::pointer> const& peers);
+
+    // Stop the recording phase of a survey and enter the reporting phase.
+    // Ignores request if no survey is active or if nonce does not match the
+    // active survey. Returns `true` if this successfully stops a survey.
+    bool stopSurveyRecording(TimeSlicedSurveyStopRecordingMessage const& msg,
```

Comment:
Thank you for making the new interfaces so clean, thought-out and well documented!